mortar/src/main/java/mortar/MortarScope.java

```diff
@@ -67,8 +66,47 @@
    * registration list. Recursively destroys all children. Parent scope drops its reference
    * to this instance. Redundant calls to this method are safe.
    */
-  void destroyChild(MortarScope child);
+  void destroy();
 
   /** Returns true if this scope has been destroyed, false otherwise. */
   boolean isDestroyed();
+
+  final class Builder {
+    private final String name;
+    private final RealScope parent;
+    private final Map<String, Object> serviceProviders = new LinkedHashMap<>();
+
+    public static Builder ofRoot() {
+      return new Builder(ROOT_NAME, null);
+    }
+
+    Builder(String name, RealScope parent) {
+      this.name = name;
+      this.parent = parent;
+    }
+
+    public Builder withService(String serviceName, Object service) {
+      Object existing = serviceProviders.put(serviceName, service);
+      if (existing != null) {
+        throw new IllegalArgumentException(
+            format("New scope \"%s\" already bound to service %s, cannot be rebound to %s", name,
+                existing, service));
+      }
+      return this;
+    }
+
+    public MortarScope build() {
+      RealScope newScope = new RealScope(name, parent, serviceProviders);
+      if (parent != null) {
+        parent.children.put(name, newScope);
+      }
+      return newScope;
```

Comment:
serviceProviders is mutable. We should either make a copy, or prevent the builder from being used again. Which usually isn't what a builder pattern is about (although here, the builder has a fixed name so it can only be used for one scope.. so not really reusable). Anyway, let's prevent bad usage.